### PR TITLE
Fix BaseModal

### PR DESCRIPTION
**Pull request by hayata-suenaga**
### Fixed Issues
$ https://github.com/Expensify/App/issues/25998
PROPOSAL: https://github.com/Expensify/App/issues/25998#issuecomment-1694423931

PR that causes this regression: https://github.com/Expensify/App/pull/25672#discussion_r1306612030

### Tests
1. Open any component that has in its' implementation [BaseModal](https://github.com/Expensify/App/blob/main/src/components/Modal/BaseModal.js)s
2. Make sure it works

#### Test 1
1. Open the app on the web
2. Open App ->Settings -> Workspace -> bank Account -> Connect manually
3. Enter test routing number and account number and proceed to step 2.
4. Click State -> type state name
5. Make sure that the right hand side panel doesn't close while you're tying

#### Test 2
1. Open the app on iOS
2. Go to settings > Workspace. Select any workspace.
3. Click on three dots icon in top right.
4. Click on delete workspace.
5. Confirm that the confirmation modal opens up asking the user to confirm if they want to delete the workspace.

#### Test 3
1. Open the app on iOS
2. Click on FAB
3. Click on Request Money
4. Check the request money page shows up

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
Yea I'm still seeing the issue

https://github.com/Expensify/App/assets/98560306/4cbbfa28-cd12-4bea-bf41-4d83c3ab95ef

**Comment by hayata-suenaga** [comment]
@allroundexperts do you have an idea what might be an issue?

**Comment by Pujan92** [comment]
If you are referencing my proposal then I have proposed below 2 changes

1. Need to add dependency `onModalHide` for the function `hideModal`
https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Modal/BaseModal.js#L75

2.  Create new `useEffect` without any dependency for unmounting logic and remove the cleanup from the existing `useEffect`.
```
useEffect(() => {
        return () => {
            // Only trigger onClose and setModalVisibility if the modal is unmounting while visible.
            if (isVisible) {
                hideModal(true);
                Modal.willAlertModalBecomeVisible(false);
            }

            // To prevent closing any modal already unmounted when this modal still remains as visible state
            Modal.setCloseModal(null);
        };
    }, []);
```

**Comment by hayata-suenaga** [comment]
I made suggestions made by @Pujan92 

Although for new chat and new room, the transition is very slow, for other items, the issue seems to be fixed

https://github.com/Expensify/App/assets/98560306/9dcb9ad4-7077-4346-bb05-bdc3a7c8ab3d

**Comment by hayata-suenaga** [comment]
@Pujan92 

I see [this comment](https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Modal/BaseModal.js#L74)

**Comment by hayata-suenaga** [comment]
By the way, I have no idea how the modal system works in NewDot...

**Comment by hayata-suenaga** [comment]
I like clicking wrong buttons 😭

**Comment by Pujan92** [comment]
> I see [this comment](https://github.com/Expensify/App/blob/c029dc509d03fd13ece058ba89ab25c4318a9050/src/components/Modal/BaseModal.js#L74)

Oh, I just read but as per the current situation I think we do need to add  `onModalHide` bcoz we are not using prop `shouldSetModalVisibility` for all modals.

**Comment by Pujan92** [comment]
> Although for new chat and new room, the transition is very slow, for other items, the issue seems to be fixed

Somehow I am not seeing slow transition compare to your video

https://github.com/Expensify/App/assets/14358475/d948c588-b076-49c4-acd4-573fb66b1e38

File: src/components/Modal/BaseModal.js [hunk 1]
```diff
@@ -90,8 +92,9 @@ function BaseModal({
 
             // To prevent closing any modal already unmounted when this modal still remains as visible state
             Modal.setCloseModal(null);
-        };
-    }, [hideModal, isVisible, onClose]);
+        },
+        [hideModal, isVisible],
```
Comment (Pujan92) [inline src/components/Modal/BaseModal.js:96]: We can remove these dependencies as not required for cleanup.

**Comment by hayata-suenaga** [comment]
@allroundexperts I pushed my changes

could you also check if the issue reported here is also fixed? My dev environment is not working correctly somehow...

File: src/components/Modal/BaseModal.js (hunk 1 above)
Comment (hayata-suenaga) [reply src/components/Modal/BaseModal.js:96]: yep I believe that's what I have done. @Pujan92 could you check the latest code?

Comment (Pujan92) [reply src/components/Modal/BaseModal.js:96]: I am still seeing the dependencies, hope you have committed. Can you plz check once

**Comment by hayata-suenaga** [comment]
I think the original issue hasn't been fixed 😢 

https://github.com/Expensify/App/assets/98560306/e778b540-ed49-4f17-852f-7214a6eb785a

File: src/components/Modal/BaseModal.js (hunk 1 above)
Comment (hayata-suenaga) [reply src/components/Modal/BaseModal.js]: ah I need to move all dependencies? didn't know that fixed!

**Comment by hayata-suenaga** [comment]
@Pujan92 by the way, I assigned you to the original issue so that you can get the payout for your help 👍 

can you help us until we get this PR merged?

**Comment by Pujan92** [comment]
> @Pujan92 by the way, I assigned you to the original issue so that you can get the payout for your help 👍

Thanks :)

> can you help us until we get this PR merged?

Sure 👍

File: src/components/Modal/BaseModal.js [hunk 2]
```diff
@@ -82,19 +82,16 @@ function BaseModal({
...
+        }
 
-            // To prevent closing any modal already unmounted when this modal still remains as visible state
-            Modal.setCloseModal(null);
-        },
-        [hideModal, isVisible],
-    );
+        // To prevent closing any modal already unmounted when this modal still remains as visible state
+        Modal.setCloseModal(null);
+    });
```
[the remark this answers is not in the document]
Comment (hayata-suenaga) [reply src/components/Modal/BaseModal.js]: haha I just committed the fix I'm so bad 😭

**Comment by hayata-suenaga** [comment]
on desktop, it's working 😄 
https://github.com/Expensify/App/assets/98560306/e1afb5ab-f278-41c9-aa5d-696bb01469a5

**Comment by hayata-suenaga** [comment]
@allroundexperts by the way, if you try the bank account flow (from the original issue), are you experiencing these flickering (hope this is not from my PR...)

https://github.com/Expensify/App/assets/98560306/24375a97-dbb2-4d8a-abfd-e2afadd9d1d2

**Comment by hayata-suenaga** [comment]
> @allroundexperts by the way, if you try the bank account flow (from the original issue), are you experiencing these flickering (hope this is not from my PR...)

okay nice this is probably something to do with my development setup then 😞 anyway, I gonna rely on you for the screenshots/recordings 🙇

**Comment by Pujan92** [comment]
Sure @allroundexperts

**Comment by hayata-suenaga** [comment]
will be back in 45min 🙇

**Review by allroundexperts — APPROVED** [review]
Waiting for @Pujan92 to upload the screenshots and a minor NAB comment. Looks good otherwise 🚀 🚀 🚀

**Comment by Pujan92** [comment]
### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/14358475/9dc6cfe4-47e2-413d-bb32-7b515cdae633

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/14358475/c45e7e04-2e27-438f-b547-f55623709d49


</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/14358475/12a6fb57-7349-4ab4-8fa7-1b2c8742c5d7

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/14358475/33e66e45-7eec-4492-ac9a-65c0d18d5584

https://github.com/Expensify/App/assets/14358475/bf8fe609-0006-4184-8232-faa0c56ebdc8


</details>

Android is taking longer to build so not added those yet. Rest are done and seems fine.

**Comment by hayata-suenaga** [comment]
@Pujan92 thank you so much for your help and testing

I requested an internal review (which is needed for PR merge) thank you for quick review @allroundexperts

**Review by robertjchen — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
Updated the QA test steps to account for all reported cases.

**Pull request merged**